File: src/main/node/pacman.js
"use strict";

const nextLocation = require("./direction.js").nextLocation;
const RIGHT = require("./direction.js").Direction.RIGHT;
const LEFT = require("./direction.js").Direction.LEFT;
const UP = require("./direction.js").Direction.UP;
const DOWN = require("./direction.js").Direction.DOWN;
const Colour = require("./colour.js");
const GameElement = require("./game-elements.js");
const Pill = require("./pills.js");
const Wall = require("./walls.js");

const icons = [
  {icon:["*","*"],direction:-1,alive:false,colour:Colour.YELLOW},
  {icon:[">","}"],direction:LEFT,alive:true,colour:Colour.YELLOW},
  {icon:["V","v"],direction:UP,alive:true,colour:Colour.YELLOW},
  {icon:["<","{"],direction:RIGHT,alive:true,colour:Colour.YELLOW},
  {icon:["Λ","^"],direction:DOWN,alive:true,colour:Colour.YELLOW}
];

module.exports =
class Pacman  extends GameElement {
  constructor (options) {
    super(options);
    if (options) {
      this.start = options.coords;
      this.game = null;
      this.playfield = null;
      this.facing = this.getDirectionByIcon(options.icon);
      this.alive = this.getLivenessByIcon(options.icon);
      this.frame = 0;    
    }
  } 
  
  static isPacman(token) {
    return (icons.filter((element) => element.icon.indexOf(token)>=0).length>0);
  }
  
  getElement(coords, icon) {
    if (Pacman.isPacman(icon)) {
      return new Pacman({coords, icon});
    }
  }
  
  addToGame(game) {
    game.addPacman(this);
  }
  
  setGame(game) {
    this.game = game;
  }
  
  setPlayfield(playfield) {
    this.playfield = playfield;
  }
  
  setDirection(facing) {
    this.facing = facing;
    let tokens = icons.find((a) => {return a.direction === facing;});
    this.image = tokens[this.frame];
  }
  
  getDirectionByIcon(icon) {
    let token = icons.find((a) => { return a.icon.indexOf(icon)>=0;});
    if (token) {
      return token.direction;
    }
  }
  
  getLivenessByIcon(icon) {
    let token = icons.find((a) => { return a.icon[0] === icon;});
    if (token) {
      return token.alive;
    }
  }
  
  colour() {
    return this.getColour(this.alive);
  }
  
  getColour(alive) {
    return icons.find((a) => { return a.alive === alive;}).colour;
  }
  
  getIconByLiveness(alive) {
    let token= icons.find((a) => { return a.alive === alive;});
    if (token) {
      return token.icon[this.frame];
    }
  }
  
  setLocation(loc) {
    loc = this.wrapLocation(loc);
    this.x = loc.x;
    this.y = loc.y;
  }
  
  icon() {
    if (this.alive) {
      return icons.find((a) => { return a.direction === this.direction();}).icon[this.frame];
    }
    return this.getIconByLiveness(this.alive);
  }
  
  animate() {
    this.frame = (this.frame+1) % 2;  
  }
  
  useAnimation() {
    this.animated = true;
  }
  
  direction () {
    return this.facing;
  }
  
  direct (direction) {
    let next = nextLocation(this.getLocation(),direction);
    if (this.isClear(next)) {
      this.setDirection(direction);
    }
  }
  
  tick() {
    let next = nextLocation(this.getLocation(),this.direction());
    if (this.isClear(next))  {
      this.setLocation(next);
      if (this.animated) {
        this.animate();
      }
    }
    if (this.isOnPill()) {
      this.game.eatPill(this.playfield.getLocation(this.getLocation()));
    }
    if (this.game.isGhost(this.getLocation())) {
      this.playfield.getLocation(this.getLocation()).collision(this);
    }
  }
  
  kill() {
    this.alive = false;
    this.image = this.getIconByLiveness(false);
    this.game.loseLife();
  }
  
  restart() {
    this.setLocation(this.start);
    this.alive=true;
  }
  
  wrapLocation(loc) {
    let width = this.playfield.width();
    let height = this.playfield.height();
    return {x:(loc.x + width) % width,y:(loc.y + height) % height};
  }
  
  isClear(loc) {
    loc = this.wrapLocation(loc);
    let element = this.playfield.getLocation(loc);
    if (!(this.game.isWall(loc))) {
      return true;
    } else if (element.isForceField()) {
      return true;
    } else {
      return false;
    }
  }
  
  isOnPill() {
    return (this.playfield.getLocation(this.getLocation()) instanceof Pill);
  }
  
  isAlive() {
    return this.alive;
  }
}

File: src/main/python/pacman.py
from direction import next_location, wrap
from direction import Direction
from game_element import GameElement
from colour import Colour


class Pacman(GameElement):
    icons = {
            "<": Direction.RIGHT,
            ">": Direction.LEFT,
            "V": Direction.UP,
            "Λ": Direction.DOWN,
            "*": "DEAD"
            }

    frame = {
            Direction.RIGHT: ["<", "{"],
            Direction.LEFT: [">", "}"],
            Direction.UP: ["V", "v"],
            Direction.DOWN: ["Λ", "^"]
            }

    def is_pacman(icon):
        if isinstance(icon, Pacman) or icon in Pacman.icons:
            return True
        return False

    def get_element(coords, icon):
        if (Pacman.is_pacman(icon)):
            return Pacman(coords, icon)

    def add_to_game(self, game):
        self.game = game
        self.width = game.field.width()
        self.height = game.field.height()
        self.use_animation(game.animation)
        game.add_pacman(self)

    def __init__(self, coordinates, icon=""):
        super(Pacman, self).__init__(coordinates, icon)
        self.start = coordinates
        self.facing = None
        if icon in Pacman.icons:
            if Pacman.icons[icon] == "DEAD":
                self.alive = False
            else:
                self.facing = Pacman.icons[icon]
                self.alive = True
        self.icon = icon
        self.frame = 0
        self.colour = Colour.YELLOW

    def set_direction(self, facing):
        if isinstance(facing, str):
            for name, item in Direction.__members__.items():
                if name == facing:
                    self.facing = item
        else:
            self.facing = facing
        self.icon = Pacman.frame[self.facing][self.frame]

    def tick(self):
        if self.clear(self.next_move(self.facing)):
            self.coordinates = (self.next_move(self.facing))
            self.animate()
        self.check_collisions()

    def animate(self):
        if self.animated:
            self.frame = (self.frame + 1) % 2
        self.icon = Pacman.frame[self.facing][self.frame]

    def check_collisions(self):
        if self.game.is_pill(self.coordinates):
            self.game.eat_pill(self.coordinates)
        elif self.game.is_ghost(self.coordinates):
            self.game.kill_ghost(self.coordinates)

    def clear(self, coordinates):
        if not self.game.is_wall(coordinates):
            return True
        if self.game.is_field(coordinates):
            return True
        return False

    def move(self, direction):
        if self.clear(self.next_move(direction)):
            self.set_direction(direction)

    def next_move(self, facing):
        return wrap(next_location(self.coordinates, facing),
                    self.width,
                    self.height)

    def kill(self):
        self.alive = False
        for icon, state in Pacman.icons.items():
            if state == "DEAD":
                self.icon = icon

    def restart(self):
        self.coordinates = self.start
        self.alive = True
        self.set_direction(self.facing)

    def use_animation(self, animated):
        self.animated = animated
